### PR TITLE
feat: add shared FastMCP 4 HTTP transport

### PR DESCRIPTION
**Pull request by glide-the**
## Summary

- serve stdio and Streamable HTTP from the same FastMCP 4 application, tool registry, instructions, version, and `ComfyCliClient` boundary
- add the `comfy-mcp serve` HTTP entry point with SDK-managed Streamable HTTP/ASGI lifecycle and loopback-safe listener configuration
- implement opt-in failure logging through a small observer boundary shared by both transports
- reorganize the oversized server module into focused internal, application, tool-registration, configuration, and transport modules without changing the thin comfy-cli passthrough rule
- document client setup, smoke testing, failure-log diagnostics, and contributing workflows, including Claude Code's variadic `--env` ordering requirement

## Public AutoDL test image

A public, reproducible AutoDL image is available for reviewing this PR in a hosted GPU environment:

[ComfyMCP AutoDL marketplace image](https://www.autodl.art/app/market/278?v=822)

- Launch: [Comfy-MCP on AutoDL](https://www.autodl.art/app/market/278?v=822)
- The image bundles ComfyUI, comfy-cli, and this PR's comfy-mcp.
- AutoDL exposes ComfyUI as application service `6006` and the Streamable HTTP MCP listener as application service `6008`; users open both from the instance page, with no SSH tunnel required.
- The deployment exercises `COMFY_MCP_PUBLIC_URL` behind AutoDL's reverse proxy so upload and download capability URLs use the complete client-facing public origin.
- This is a reviewer test artifact, not a runtime dependency or a project-owned hosted service.

## Testing

- `pytest -q` — 2272 passed, 10 deselected, both locally and on the AutoDL image
- `ruff check .`
- `ruff format --check .`
- automated stdio and Streamable HTTP discovery: the same 40-tool application after merging the current upstream `main`
- local stdio MCP business flow: initialize, discover the shared tool surface, and call `server_info` against comfy-cli 1.16.0
- isolated Claude Code stdio registration smoke test with two environment variables
- live AutoDL deployment: initialize through the public reverse proxy, discover all 40 tools including `billing_status`, confirm the running ComfyUI through `server_info`, and verify that an HTTP upload capability URL preserves HTTPS, the public host, external port `8443`, and the same-listener `/api/uploads/` route

### CODE REVIEW

**Review by SiaoZeng — CHANGES_REQUESTED** [review]
please dismiss my suggestions.

**Comment by glide-the** [comment]
I have read and agree to the Contributor License Agreement

**Comment by glide-the** [comment]
Review follow-up for `ed05c6e`:

- The CLI docs now spell the two valid forms separately as `comfy --json --where local …` and `comfy --json-stream --where local …`.
- Upload and download capability URLs now use `_UPLOAD_ROUTE` and `_DOWNLOAD_ROUTE` as their single source of truth.
- The Windows path assertion now uses `relative_to(...).as_posix()`.
- The stdio subprocess flow skips non-POSIX systems and has a 120-second total timeout.
- Approval caching now uses explicit gate IDs (`spend_generate` and `spend_optin`). `COMFY_MCP_ASSUME_CONSENT` remains unable to authorize either spend gate.
- Modern protocol detection now checks membership in `MODERN_PROTOCOL_VERSIONS`, and the tests use the package's `LATEST_MODERN_VERSION` / `LATEST_HANDSHAKE_VERSION` constants.
- Signed download lease release now runs in `finally`; a regression test forces a body-send failure and verifies scratch cleanup.

On the `Forwarded` / `X-Forwarded-*` finding: this behavior is intentional and follows the remote filesystem contract. In HTTP mode, `file_path` and `out_dir` are client-local, so transfer commands must use the public origin through which the current client reached the MCP server. Behind a reverse proxy, the ASGI request URL contains only the internal upstream address; the forwarded headers are therefore the source of the client-facing scheme, host, and port.

The deployment contract requires the reverse proxy to sanitize and replace these headers and to protect `/mcp`, `/api/uploads/`, and `/downloads/`. Platforms that do not preserve the authoritative origin use `COMFY_MCP_PUBLIC_URL`.

A caller changing these headers only changes the capability command returned to that same caller; it does not redirect another client or grant access to another client's files. Adding another application-level proxy-trust switch would duplicate the proxy's trust policy and break the default contract that resources belong to the client environment where the user invoked the tool. No change is planned for this item.

File: tests/test_fastmcp_app.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        self, *args: str, timeout=None, plain_ok=False, stdout_cap=None
+    ):
+        self.calls.append(args)
+        return {"ok": True}
+
+    async def run_streaming(
+        self, *args: str, ctx=None, timeout=None, raise_on_timeout=True
+    ):
+        self.calls.append(args)
+        return {"prompt_id": "prompt-1", "status": "completed"}
```
[the remark this answers is not in the document]
Comment (glide-the) [reply tests/test_fastmcp_app.py]: Declining this refactor. `_FlowClient` is intentionally module-local: this is the only test that binds a `ComfyCliClient` port implementation through `client_context.bind_client()`. The shared `patched_run`, `patched_stream`, and `patched_async_run` fixtures mock the concrete subprocess spawn boundary for runner/tool tests; they do not duplicate `_FlowClient`, and making them depend on an application-port fake would conflate two distinct test layers. Moving a one-consumer fake into `conftest.py` would broaden the shared fixture surface without creating a single source of truth or removing duplication. We can extract it if a second port-level consumer appears.

File: tests/test_stdio_business_flow.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            "local_path": command[1],
+            "cloud_name": "stdio-input.bin",
+        }}],
+    }})
+elif command == ["run", "--workflow", "/srv/workflows/smoke.json"]:
+    envelope({{"prompt_id": "prompt-stdio"}})
+elif command == ["jobs", "status", "prompt-stdio"]:
+    envelope({{"prompt_id": "prompt-stdio", "status": "completed"}})
+elif command == ["download", "prompt-stdio", "-o", "/tmp/results"]:
+    envelope({{"files": [{{"path": "/tmp/results/result.png"}}]}})
```
[the remark this answers is not in the document]
Comment (glide-the) [reply tests/test_stdio_business_flow.py]: Fixed in dbbb34e. The mocked stdio-process test was removed and the process flow now lives in the opt-in `tests/e2e/test_smoke.py` suite. The E2E derives its destination from pytest's per-test directory (`tmp_path / "stdio_out"`), so concurrent runs no longer share or create `/tmp/results`.

File: tests/test_remote_http.py [hunk 3]
```diff
@@ -0,0 +1,784 @@
...
+    assert status_calls[0]["cmd"][4:] == ["jobs", "status", "prompt-http"]
+    assert output_calls[0]["cmd"][4:] == [
+        "download",
+        "prompt-http",
+        "-o",
+        download_scratch[0],
+    ]
+    assert pathlib.Path(download_scratch[0]).exists()
+    file_transfer._DOWNLOAD_STORE.close()
+    assert not pathlib.Path(download_scratch[0]).exists()
```
[the remark this answers is not in the document]
Comment (glide-the) [reply tests/test_remote_http.py]: Fixed in dbbb34e. `tests/conftest.py` now owns transfer capability cleanup in an autouse `_clear_file_capabilities` fixture. Its finalizer closes both `_UPLOAD_STORE` and `_DOWNLOAD_STORE` after every test, including assertion failures, and the in-test `close()` calls were removed.

File: AGENTS.md [hunk 4]
```diff
@@ -8,24 +8,27 @@ standalone [MCP](https://modelcontextprotocol.io) server that lets an agent driv
...
-Every tool is a passthrough to the `comfy` binary — the only way to reach comfy-cli is the
-`_run_comfy(*args)` helper in `src/comfy_mcp/server.py`, which shells out to `comfy --json
---where local <args>` (global flags **before** the subcommand), parses comfy-cli's versioned
-`envelope/1` result, and returns its `data`. Do not bypass it.
+Every tool is a passthrough to the `comfy` binary. Application/tool code reaches it only
+through the `ComfyCliClient` port under `src/comfy_mcp/client/`; the concrete client is
+composed with the guarded `_run_comfy*_impl` runners in `server/_internal.py`. Those runners shell out
+to `comfy --json[--stream] --where local <args>` (global flags **before** the subcommand),
+parse comfy-cli's versioned `envelope/1`, and return its data. Do not bypass the client port
+or the runners.
```
[the remark this answers is not in the document]
Comment (glide-the) [reply AGENTS.md]: Fixed in `ed05c6e`. `AGENTS.md` now documents the two real runner forms separately: `comfy --json --where local ...` and `comfy --json-stream --where local ...`. The invalid combined spelling is no longer presented.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -17,24 +18,62 @@ response-time SLA.
...
-**Every tool is a passthrough to the `comfy` binary.** There is exactly one way
-to reach comfy-cli — the `_run_comfy(*args)` helper in
-`src/comfy_mcp/server.py`, which shells out to
-`comfy --json --where local <args>`, parses comfy-cli's versioned `envelope/1`
-result, and returns its `data`. Do not bypass it.
+**Every tool is a passthrough to the `comfy` binary.** Application functions
+resolve a `ComfyCliClient` from `src/comfy_mcp/client/`; its concrete
+implementation delegates to the established guarded runners in
+`src/comfy_mcp/server/_internal.py`. Those runners are the only code that may spawn
+`comfy --json[--stream] --where local <args>`, parse comfy-cli's versioned
```
[the remark this answers is not in the document]
Comment (glide-the) [reply CONTRIBUTING.md]: Fixed in `ed05c6e`. `CONTRIBUTING.md` now lists `--json` and `--json-stream` as separate global flags, so the documentation cannot be read as the nonexistent `--json--stream` option.

File: SECURITY.md [hunk 6]
```diff
@@ -44,19 +44,20 @@ This section says what this server assumes, what it deliberately allows, and wha
...
-- **Resource exhaustion of the server itself.** Timeouts are bounded with hard ceilings (`_bounded_timeout`), output drains are capped to a trailing window (`_drain_capped_async`), a timeout kills the whole process tree rather than orphaning a `git pull` or a multi-GB transfer (`_kill_proc_tree`), and inline image previews are capped at 8 files / 16 MiB aggregate so a large batch cannot blow up a reply.
+- **Remote file-transfer containment.** Upload URLs bind an unguessable, single-use token to one validated image filename, expire after five minutes, and accept at most 50 MiB. The route writes owner-only scratch, passes only that generated path to comfy-cli, never records request bytes in the failure log, and removes scratch on every exit. Signed output URLs bind an unguessable token, filename, and expiry with HMAC; only regular files that comfy-cli reported inside the dedicated download scratch root can be published. Responses are non-cacheable and scratch is removed at expiry. Treat complete capability URLs as temporary secrets and configure a reverse proxy to protect `/mcp`, `/api/uploads/`, and `/downloads/`.
```
[the remark this answers is not in the document]
Comment (glide-the) [reply SECURITY.md:59]: Fixed in `ed05c6e`. Signed downloads now use `_LeasedFileResponse`, whose ASGI `__call__` releases the lease in a `finally` block. A send failure or cancellation can no longer bypass release, so expiry or shutdown can remove the scratch directory.

File: src/comfy_mcp/file_transfer.py [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+                key: value
+                for key, value in self._entries.items()
+                if value.expires >= now
+            }
+            self._entries[token] = _UploadEntry(
+                request.filename,
+                request.mime_type,
+                now + self._ttl_seconds,
+            )
+        url = f"{base_url.rstrip('/')}/api/uploads/{token}"
```
[the remark this answers is not in the document]
Comment (glide-the) [reply src/comfy_mcp/file_transfer.py]: Fixed in `ed05c6e`. Upload and download capability URLs are now formatted from `_UPLOAD_ROUTE` and `_DOWNLOAD_ROUTE`; the registered route constants are the single source of truth.

File: src/comfy_mcp/file_transfer.py [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+                {"error": "file exceeds the 50 MiB upload limit"},
+                status_code=413,
+                headers={"Cache-Control": "no-store"},
+            )
+        chunks.append(chunk)
+
+    directory = tempfile.TemporaryDirectory(prefix="comfy-mcp-upload-")
+    path = os.path.join(directory.name, entry.filename)
+    try:
+        await asyncio.to_thread(_write_owner_only, path, b"".join(chunks))
```
[the remark this answers is not in the document]
Comment (glide-the) [reply src/comfy_mcp/file_transfer.py]: Fixed in `dbbb34e`. `receive_upload` opens the owner-only scratch file before consuming the body and writes each incoming chunk directly to it. The running 50 MiB limit is still enforced without retaining the complete request body in memory.

File: src/comfy_mcp/file_transfer.py [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
+    return FileResponse(
+        lease.entry.path,
+        filename=lease.entry.filename,
+        headers={
+            "Cache-Control": "private, no-store",
+            "Content-Security-Policy": "default-src 'none'",
+            "X-Content-Type-Options": "nosniff",
+        },
+        background=BackgroundTask(_DOWNLOAD_STORE.release, lease),
+    )
```
[the remark this answers is not in the document]
Comment (glide-the) [reply src/comfy_mcp/file_transfer.py]: Fixed in `ed05c6e`. Lease release moved from a Starlette background task into `_LeasedFileResponse.__call__` under `finally`. The regression test forces body delivery to fail and verifies that closing the store can remove the scratch directory.

File: src/comfy_mcp/server/_internal.py [hunk 10]
```diff
@@ -4254,6 +3924,44 @@ class SpendApproval(BaseModel):
...
+    """Whether FastMCP negotiated the sessionless 2026 protocol era."""
+
+    if ctx is None:
+        return False
+    try:
+        request_context = getattr(ctx, "request_context")
+        version = getattr(request_context, "protocol_version")
+    except (AttributeError, RuntimeError, ValueError):
+        return False
+    return str(version) >= str(types.LATEST_PROTOCOL_VERSION)
```
[the remark this answers is not in the document]
Comment (glide-the) [reply src/comfy_mcp/server/_internal.py]: Fixed in `ed05c6e` and refined in `98d9508`. `_is_modern_protocol` now tests explicit membership in `MODERN_PROTOCOL_VERSIONS`; negotiated modern, handshake, stale, pre-initialize, and unknown versions are covered without lexicographic comparison.

File: src/comfy_mcp/server/_internal.py [hunk 11]
```diff
@@ -4395,27 +4110,68 @@ class _ApprovalWording(NamedTuple):
...
-# `escape_hatch` on the RAISE paths already covers the client that cannot prompt.
-_DECLINE_MAY_BE_AUTOMATIC = (
-    ' Some clients answer "decline" on their own when they cannot display a '
-    "prompt, so if no prompt appeared, the client refused by itself and nobody "
-    "was asked."
-)
+def _approval_key(wording: _ApprovalWording) -> str:
+    """Stable key shared by every round of one approval gate."""
+
+    return f"comfy_mcp_{wording.consent_token or 'spend'}"
```
[the remark this answers is not in the document]
Comment (glide-the) [reply src/comfy_mcp/server/_internal.py]: Fixed in `ed05c6e`. `_ApprovalWording.gate_id` now gives the two spend flows distinct approval keys, `spend_generate` and `spend_optin`. Both retain an empty `consent_token`, so `COMFY_MCP_ASSUME_CONSENT` cannot authorize either spend gate.

File: src/comfy_mcp/server/_internal.py [hunk 12]
```diff
@@ -6319,30 +6051,151 @@ def _select_inline_images(paths: list[str]) -> list[str]:
...
+    if (
+        not _host_has_port(host)
+        and forwarded_port.isascii()
+        and forwarded_port.isdigit()
+    ):
+        port = int(forwarded_port)
+        if 1 <= port <= 65535:
+            host = f"{host}:{port}"
+
+    return f"{scheme}://{host}"
```
[the remark this answers is not in the document]
Comment (glide-the) [reply src/comfy_mcp/server/_internal.py:6754]: Declining this change. The HTTP filesystem adapter follows the convention that `file_path` and `out_dir` belong to the environment where the user invoked the tool, so capability commands must use the client-facing origin carried by that request. The deployment contract requires the reverse proxy to sanitize and replace `Forwarded` / `X-Forwarded-*` and protect `/mcp`, `/api/uploads/`, and `/downloads/`; deployments that cannot preserve that origin use the authoritative `COMFY_MCP_PUBLIC_URL` fallback. A caller altering these headers changes only the command returned to that same caller and does not redirect another client or grant access to another client’s files. Adding an application-level proxy-trust switch would duplicate the deployment trust policy and break this convention-over-configuration contract. No code change is planned.

File: src/comfy_mcp/server/_internal.py [hunk 13]
```diff
@@ -6354,6 +6207,44 @@ def fetch_outputs(
...
+                            published,
+                            client_os=client_os,
+                        ),
+                    )
+                ],
+                structured_content=published,
+            )
+        except BaseException:
+            directory.cleanup()
+            raise
```
[the remark this answers is not in the document]
Comment (glide-the) [reply src/comfy_mcp/server/_internal.py:6827]: Fixed in `ed05c6e`. The cleanup guard now covers only the download and publish operations. Once `publish_downloads` returns, ownership has transferred to the signed-download store and `ToolResult` construction occurs outside the cleanup scope, preventing a live token from losing its bytes.

File: src/comfy_mcp/server/_internal.py [hunk 14]
```diff
@@ -11851,6 +11340,38 @@ async def upload_file(paths: list[str], overwrite: bool = False) -> Any:
...
+    re-invoke this tool after an expiry/failure rather than reusing it. Images
+    are limited to 50 MiB. Base64/data-inline upload is not supported.
+    """
+    request = file_transfer.prepare_upload_request(file_path, client_os)
+    if _is_remote_http_request(ctx):
+        return file_transfer.mint_upload_command(
+            request,
+            base_url=_remote_http_base_url(),
+        )
+    return await _upload_one_file(request.file_path)
```
[the remark this answers is not in the document]
Comment (glide-the) [reply src/comfy_mcp/server/_internal.py:12309]: Fixed in `dbbb34e`. The `_run_comfy_async` documentation and `clitext.py` now describe the active single-file `upload_file(file_path, client_os)` contract. The intentional `fetch_outputs` behavior is unchanged.

File: tests/e2e/test_remote_mcp.py [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+import re
+import sys
+import time
+import uuid
+from typing import Any
+from urllib.parse import urlsplit
+
+import httpx2
+import pytest
+from fastmcp import Client
```
[the remark this answers is not in the document]
Comment (glide-the) [reply tests/e2e/test_remote_mcp.py]: Fixed in `dbbb34e`. The opt-in deployed E2E now loads `httpx2` through `pytest.importorskip("httpx2")`. This keeps the development/E2E dependency boundary separate from the released runtime while producing a clean skip when the extra is absent.

File: tests/e2e/test_smoke.py [hunk 16]
```diff
@@ -16,21 +16,14 @@
...
-CI runners have neither, so they SKIP cleanly there — CI stays green via skip, not
-failure. Run them on a machine that has both with::
+
+**Gated.** They require BOTH a live same-machine ComfyUI answering on
+``COMFY_LOCAL_URL`` (or ``http://127.0.0.1:8188``) AND the ``comfy`` binary on
+``PATH`` (or ``COMFY_BIN``). ``COMFYUI_URL`` / ``COMFYUI_HOST`` are deliberately
+rejected here: those configure the wrapper's remote ComfyUI routing, while this
+suite validates its local engine boundary. CI runners have neither, so they SKIP
+cleanly there — CI stays green via skip, not failure. Run them on a machine that
+has both with::
```
[the remark this answers is not in the document]
Comment (glide-the) [reply tests/e2e/test_smoke.py:40]: Declining this change. `tests/e2e/test_smoke.py` is an explicitly opt-in live integration suite, not part of the isolated default test suite. Its purpose is to validate the real `comfy` binary against a live local ComfyUI, which shared fixtures cannot exercise. Default tests continue to use the shared fixtures and the live suite skips when its prerequisites are absent. `dbbb34e` also moved real stdio-process coverage into this E2E suite and removed the generated fake executable.

File: tests/test_address_env_docs.py [hunk 17]
```diff
@@ -78,10 +81,16 @@ def test_server_reads_the_comfyui_vars_and_not_comfy_local_url():
...
     """
-    sources = {path.name: path.read_text(encoding="utf-8") for path in _PACKAGE_SOURCES}
-    assert "server.py" in sources, "the package layout moved out from under this test"
+    sources = {
+        str(path.relative_to(_PACKAGE_ROOT)): path.read_text(encoding="utf-8")
+        for path in _PACKAGE_SOURCES
+    }
+    assert "server/_internal.py" in sources, (
+        "the private server implementation moved out from under this test"
+    )
```
[the remark this answers is not in the document]
Comment (glide-the) [reply tests/test_address_env_docs.py:90]: Fixed in `ed05c6e`. The source-map keys now use `path.relative_to(_PACKAGE_ROOT).as_posix()`, keeping the assertion stable on Windows and POSIX systems.

File: tests/test_fastmcp_app.py [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+    monkeypatch,
+):
+    engine = _FlowClient()
+    monkeypatch.setattr(server, "_detect_comfy_cli_version", lambda: "1.14.0")
+
+    tools, info, submitted, status, outputs = asyncio.run(
+        _main_application_flow(engine)
+    )
+
+    assert len(tools) == 39
```
[the remark this answers is not in the document]
Comment (glide-the) [reply tests/test_fastmcp_app.py]: Fixed in `dbbb34e`. Discovery tests now compare the discovered names with the shared `EXPECTED_TOOL_NAMES = frozenset(server_tools.__all__)`; export checks assert uniqueness rather than a fixed count. Tool additions no longer require sweeping numeric updates.

File: tests/test_fastmcp_app.py [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+def test_shared_application_paid_workflow_keeps_legacy_elicitation_compatible(
+    monkeypatch,
+):
+    engine = _FlowClient()
+    monkeypatch.setattr(server, "_comfy_run_takes_allow_spend", lambda: True)
+
+    protocol, result = asyncio.run(_shared_paid_workflow("legacy", engine))
+
+    assert str(protocol) != "2026-07-28"
+    assert result.data["prompt_id"] == "prompt-1"
```
[the remark this answers is not in the document]
Comment (glide-the) [reply tests/test_fastmcp_app.py]: Addressed in `ed05c6e` and completed in `98d9508`. The tests assert `LATEST_MODERN_VERSION` and `LATEST_HANDSHAKE_VERSION` from the pinned `mcp_types.version` package and cover the exact modern/legacy negotiation boundary.

File: tests/test_file_transfer.py [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+    query = parse_qs(parsed.query)
+    expires = int(query["expires"][0])
+    signature = query["signature"][0]
+
+    lease = store.acquire(token, unquote(filename), expires, signature)
+    assert lease is not None
+    store.release(lease)
+    monkeypatch.setattr(file_transfer.time, "time", lambda: now + 11)
+    assert store.acquire(token, unquote(filename), expires, signature) is None
+    assert not scratch.exists()
```
[the remark this answers is not in the document]
Comment (glide-the) [reply tests/test_file_transfer.py:256]: No change is needed for this test. It drives the batch through expiry before returning: `acquire` invokes `_expire`, which removes the entries and cleans the scratch directory synchronously. The later daemon callback exits immediately because `batch.expired` is already true. The suggested trailing `store.close()` would not cancel that timer because the expired batch is no longer present in `_entries`, so it would add no cleanup guarantee or test isolation.

File: tests/test_remote_http.py [hunk 21]
```diff
@@ -0,0 +1,784 @@
...
+    finally:
+        if proc.poll() is None:
+            proc.kill()
+            proc.communicate()
+
+    # Uvicorn 0.52 completes the ASGI lifespan, restores the process's prior
+    # signal handler, then replays the signal. Python's SIGINT handler becomes
+    # KeyboardInterrupt (normalized by ``serve``); SIGTERM keeps native Unix
+    # signal-exit semantics. Both paths must finish application shutdown first.
+    assert proc.returncode == expected_returncode
```
[the remark this answers is not in the document]
Comment (glide-the) [reply tests/test_remote_http.py]: Fixed in `dbbb34e`. The unsupported Uvicorn version number was removed while preserving the explanation of lifespan completion, signal-handler restoration, signal replay, and the expected SIGINT/SIGTERM exit semantics.

File: tests/test_stdio_business_flow.py [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+else:
+    print("unexpected command: " + repr(command), file=sys.stderr)
+    raise SystemExit(3)
+"""
```
[the remark this answers is not in the document]
Comment (glide-the) [reply tests/test_stdio_business_flow.py]: Fixed in `dbbb34e`. The generated-executable stdio test was removed from the default suite. Real stdio-process coverage now lives in the opt-in `tests/e2e/test_smoke.py`, while default tests use the shared fixtures in `tests/conftest.py`.

File: tests/test_stdio_business_flow.py [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+def test_real_stdio_process_completes_submit_poll_fetch(tmp_path):
+    fake = tmp_path / "fake-comfy"
+    fake.write_text(_FAKE_COMFY.format(python=sys.executable), encoding="utf-8")
+    fake.chmod(0o755)
+    input_path = tmp_path / "stdio-input.png"
+    input_path.write_bytes(b"stdio-input")
+
+    tools, info, uploaded, submitted, status, outputs = asyncio.run(
+        _run_stdio_flow(str(fake), str(input_path))
+    )
```
[the remark this answers is not in the document]
Comment (glide-the) [reply tests/test_stdio_business_flow.py]: Addressed in `ed05c6e` and superseded by `dbbb34e`. The original generated POSIX executable was removed entirely. The opt-in live stdio flow uses `sys.executable`, the real `comfy` binary, and an overall `asyncio.wait_for` deadline, so it has neither the shebang/executable-bit dependency nor an unbounded wait.

File: AGENTS.md [hunk 24]
```diff
@@ -147,14 +237,38 @@ Tests live in `tests/` and mock comfy-cli — no real ComfyUI, no `comfy` binary
 and the parser are exercised directly (`test_wrapper.py`, `test_parser.py`); each tool group
 has its own file. Add a tool's test with it.
 
+Remote configuration and same-instance adapter tests live in `test_remote_transport.py`;
+in-process FastMCP business flows live in `test_fastmcp_app.py`; real-loopback Streamable HTTP
+integration lives in `test_remote_http.py`; transfer edge cases live in `test_file_transfer.py`.
+The real stdio process flow belongs to the opt-in `e2e/test_smoke.py` suite and uses the real
```
[the remark this answers is not in the document]
Comment (glide-the) [reply AGENTS.md]: Fixed in `717e9b4`. `AGENTS.md` now uses the repository-relative path `tests/e2e/test_smoke.py`.

**Comment by glide-the** [comment]
功能差异过大，需要后续整理出功能对比清单
